File: src/calculus_volumes_of_revolution_easy.py
"""
calculus - volumes_of_revolution (easy)
Generated: 2026-02-22T05:17:11.456789
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 4)

    if problem_type == 1:
        # Disk method: revolve y = h (constant) around x-axis from 0 to b
        height = randint(1, 8)
        b_bound = randint(1, 8)

        volume = pi * height**2 * b_bound

        question = (f"Find the volume when the region bounded by $y = {height}$, $y = 0$, "
                    f"$x = 0$, and $x = {b_bound}$ is revolved around the $x$-axis.")

        solution = steps(
            f"Using the disk method: $V = \\pi \\int_{{0}}^{{{b_bound}}} [{height}]^2 \\, dx$",
            f"$V = \\pi \\cdot {height**2} \\int_{{0}}^{{{b_bound}}} dx$",
            f"$V = \\pi \\cdot {height**2} \\cdot {b_bound}$",
            f"$V = {latex(volume)}$"
        )

        return problem(
            question=question,
            answer=volume,
            difficulty=(1000, 1100),
            topic="calculus/volumes_of_revolution",
            solution=solution
        )

    elif problem_type == 2:
        # Disk method: revolve y = mx around x-axis from 0 to b
        m_val = randint(1, 5)
        b_bound = randint(1, 8)

        # V = π * m² * b³/3
        volume = pi * m_val**2 * b_bound**3 / 3

        func_str = f"y = {m_val}x" if m_val != 1 else "y = x"

        question = (f"Find the volume when the region bounded by ${func_str}$, $y = 0$, "
                    f"and $x = {b_bound}$ is revolved around the $x$-axis.")

        solution = steps(
            f"Using the disk method: $V = \\pi \\int_{{0}}^{{{b_bound}}} ({m_val}x)^2 \\, dx$",
            f"$V = \\pi \\cdot {m_val**2} \\int_{{0}}^{{{b_bound}}} x^2 \\, dx$",
            f"$V = \\pi \\cdot {m_val**2} \\cdot \\left[\\frac{{x^3}}{{3}}\\right]_{{0}}^{{{b_bound}}}$",
            f"$V = \\pi \\cdot {m_val**2} \\cdot \\frac{{{b_bound**3}}}{{3}}$",
            f"$V = {latex(volume)}$"
        )

        return problem(
            question=question,
            answer=volume,
            difficulty=(1100, 1250),
            topic="calculus/volumes_of_revolution",
            solution=solution
        )

    elif problem_type == 3:
        # Disk method: revolve y = √x around x-axis from 0 to b
        b_bound = randint(1, 8)

        volume = pi * b_bound**2 / 2

        question = (f"Find the volume when the region bounded by $y = \\sqrt{{x}}$, $y = 0$, "
                    f"and $x = {b_bound}$ is revolved around the $x$-axis.")

        solution = steps(
            f"Using the disk method: $V = \\pi \\int_{{0}}^{{{b_bound}}} (\\sqrt{{x}})^2 \\, dx$",
            f"$V = \\pi \\int_{{0}}^{{{b_bound}}} x \\, dx$",
            f"$V = \\pi \\left[\\frac{{x^2}}{{2}}\\right]_{{0}}^{{{b_bound}}}$",
            f"$V = \\pi \\cdot \\frac{{{b_bound**2}}}{{2}}$",
            f"$V = {latex(volume)}$"
        )

        return problem(
            question=question,
            answer=volume,
            difficulty=(1200, 1300),
            topic="calculus/volumes_of_revolution",
            solution=solution
        )

    else:
        # Disk method: revolve y = x^(1/n) around x-axis from 0 to b
        # n in {3, 4} so the exponent is 1/3 or 1/4
        n = choice([3, 4])
        b_bound = randint(1, 6)

        # y^2 = x^(2/n), ∫₀ᵇ x^(2/n) dx = b^(2/n+1)/(2/n+1)
        from sympy import Rational as Rat
        exp_sq = Rat(2, n)
        volume = pi * b_bound**(exp_sq + 1) / (exp_sq + 1)
        volume = simplify(volume)

        root_str = f"\\sqrt[{n}]{{x}}"

        question = (f"Find the volume when the region bounded by $y = {root_str}$, $y = 0$, "
                    f"and $x = {b_bound}$ is revolved around the $x$-axis.")

        solution = steps(
            f"Using the disk method: $V = \\pi \\int_{{0}}^{{{b_bound}}} \\left({root_str}\\right)^2 \\, dx$",
            f"$V = \\pi \\int_{{0}}^{{{b_bound}}} x^{{2/{n}}} \\, dx$",
            f"$V = \\pi \\left[\\frac{{x^{{2/{n}+1}}}}{{2/{n}+1}}\\right]_{{0}}^{{{b_bound}}}$",
            f"$V = {latex(volume)}$"
        )

        return problem(
            question=question,
            answer=volume,
            difficulty=(1200, 1300),
            topic="calculus/volumes_of_revolution",
            solution=solution
        )

emit(generate())

File: src/calculus_volumes_of_revolution_easy.jl
# calculus - volumes_of_revolution (easy)
# Generated: 2026-03-08T20:48:42.522513

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("calculus/volumes_of_revolution")
    
    problem_type = choice(1:4)
    
    if problem_type == 1
        # Disk method: revolve y = constant or y = mx around x-axis
        a = randint(1, 8)
        b = randint(a+1, a+6)
        
        if choice([true, false])
            # Constant function y = k
            k = randint(2, 8)
            f_expr = k
            volume = π * k^2 * (b - a)
            integrand = k^2
            
            soln = steps(
                "Revolve \$y = $(k)\$ around the x-axis from \$x = $(a)\$ to \$x = $(b)\$",
                sol("Volume formula", "V = \\pi \\int_{$(a)}^{$(b)} [f(x)]^2 \\, dx"),
                sol("Setup", "V = \\pi \\int_{$(a)}^{$(b)} $(k^2) \\, dx"),
                sol("Evaluate", "V = $(k^2)\\pi (x) \\Big|_{$(a)}^{$(b)} = $(k^2)\\pi($(b) - $(a))"),
                sol("Answer", "V = $(volume//π)\\pi")
            )
            
            ans_tex = "$(volume//π)\\pi"
        else
            # Linear function y = cx
            c = randint(1, 5)
            f_expr = c*x
            volume = π * c^2 * (b^3 - a^3) // 3
            
            soln = steps(
                "Revolve \$y = $(c)x\$ around the x-axis from \$x = $(a)\$ to \$x = $(b)\$",
                sol("Volume formula", "V = \\pi \\int_{$(a)}^{$(b)} [f(x)]^2 \\, dx"),
                sol("Setup", "V = \\pi \\int_{$(a)}^{$(b)} ($(c)x)^2 \\, dx = $(c^2)\\pi \\int_{$(a)}^{$(b)} x^2 \\, dx"),
                sol("Evaluate", "V = $(c^2)\\pi \\cdot \\frac{x^3}{3} \\Big|_{$(a)}^{$(b)} = $(c^2)\\pi \\cdot \\frac{$(b^3) - $(a^3)}{3}"),
                sol("Answer", "V = \\frac{$(c^2 * (b^3 - a^3))}{3}\\pi")
            )
            
            ans_tex = "\\frac{$(c^2 * (b^3 - a^3))}{3}\\pi"
        end
        
        problem(
            question = "Find the volume of the solid obtained by revolving the region bounded by \$y = $(tex(f_expr))\$, \$x = $(a)\$, \$x = $(b)\$, and the x-axis around the x-axis.",
            answer = ans_tex,
            difficulty = (700, 900),
            solution = soln,
            time = 120
        )
        
    elseif problem_type == 2
        # Disk method: revolve y = √x around x-axis
        a = choice([0, 1])
        b = randint(4, 16)
        
        volume = π * (b^2 - a^2) // 2
        
        soln = steps(
            "Revolve \$y = \\sqrt{x}\$ around the x-axis from \$x = $(a)\$ to \$x = $(b)\$",
            sol("Volume formula", "V = \\pi \\int_{$(a)}^{$(b)} [f(x)]^2 \\, dx"),
            sol("Setup", "V = \\pi \\int_{$(a)}^{$(b)} (\\sqrt{x})^2 \\, dx = \\pi \\int_{$(a)}^{$(b)} x \\, dx"),
            sol("Evaluate", "V = \\pi \\cdot \\frac{x^2}{2} \\Big|_{$(a)}^{$(b)} = \\pi \\cdot \\frac{$(b^2) - $(a^2)}{2}"),
            sol("Answer", "V = \\frac{$(b^2 - a^2)}{2}\\pi")
        )
        
        problem(
            question = "Find the volume of the solid obtained by revolving the region bounded by \$y = \\sqrt{x}\$, \$x = $(a)\$, \$x = $(b)\$, and the x-axis around the x-axis.",
            answer = "\\frac{$(b^2 - a^2)}{2}\\pi",
            difficulty = (800, 1000),
            solution = soln,
            time = 150
        )
        
    elseif problem_type == 3
        # Washer method: revolve region between y = c and y = mx around x-axis
        a = randint(1, 6)
        b = randint(a+2, a+8)
        k = randint(2, 7)
        m = randint(1, 3)
        
        # Outer radius: y = k, Inner radius: y = mx
        outer_vol = π * k^2 * (b - a)
        inner_vol = π * m^2 * (b^3 - a^3) // 3
        volume = outer_vol - inner_vol
        
        soln = steps(
            "Revolve region between \$y = $(k)\$ and \$y = $(m)x\$ around the x-axis from \$x = $(a)\$ to \$x = $(b)\$",
            sol("Washer method", "V = \\pi \\int_{$(a)}^{$(b)} [R(x)^2 - r(x)^2] \\, dx"),
            sol("Setup", "V = \\pi \\int_{$(a)}^{$(b)} [$(k^2) - $(m^2)x^2] \\, dx"),
            sol("Evaluate", "V = \\pi [$(k^2)x - \\frac{$(m^2)x^3}{3}] \\Big|_{$(a)}^{$(b)}"),
            sol("Answer", "V = \\pi[$(k^2 * b) - \\frac{$(m^2 * b^3)}{3} - $(k^2 * a) + \\frac{$(m^2 * a^3)}{3}]")
        )
        
        numer = 3*k^2*(b-a) - m^2*(b^3 - a^3)
        
        problem(
            question = "Find the volume of the solid obtained by revolving the region bounded by \$y = $(k)\$, \$y = $(m)x\$, \$x = $(a)\$, and \$x = $(b)\$ around the x-axis.",
            answer = "\\frac{$(numer)}{3}\\pi",
            difficulty = (900, 1100),
            solution = soln,
            time = 180
        )
        
    else
        # Shell method: revolve y = k around y-axis
        a = randint(1, 6)
        b = randint(a+2, a+8)
        k = randint(2, 8)
        
        volume = π * k * (b^2 - a^2)
        
        soln = steps(
            "Revolve \$y = $(k)\$ around the y-axis from \$x = $(a)\$ to \$x = $(b)\$",
            sol("Shell method", "V = 2\\pi \\int_{$(a)}^{$(b)} x \\cdot f(x) \\, dx"),
            sol("Setup", "V = 2\\pi \\int_{$(a)}^{$(b)} x \\cdot $(k) \\, dx = $(2*k)\\pi \\int_{$(a)}^{$(b)} x \\, dx"),
            sol("Evaluate", "V = $(2*k)\\pi \\cdot \\frac{x^2}{2} \\Big|_{$(a)}^{$(b)} = $(k)\\pi($(b^2) - $(a^2))"),
            sol("Answer", "V = $(k*(b^2 - a^2))\\pi")
        )
        
        problem(
            question = "Find the volume of the solid obtained by revolving the region bounded by \$y = $(k)\$, \$x = $(a)\$, \$x = $(b)\$, and the x-axis around the y-axis using the shell method.",
            answer = "$(k*(b^2 - a^2))\\pi",
            difficulty = (1000, 1200),
            solution = soln,
            time = 180
        )
    end
end